contracts/mocks/SmtV2Mock.sol

```diff
@@ -1,183 +1,315 @@
+     * It should be used cautiously as wrong chronological order of root history
+     * can lead to incorrect behaviour on the client side.
+     * @param _i index of node
+     * @param _v value of node
+     * @param _timestamp timestamp of root
+     * @param _blockNumber block number of root
+     */
+    function addHistorical(
+        SmtData storage self,
+        uint256 _i,
+        uint256 _v,
+        uint64 _timestamp,
+        uint64 _blockNumber
+    ) public {
+        Node memory node = Node(NodeType.LEAF, 0, 0, _i, _v);
+        self.root = addLeaf(self, node, self.root, 0);
 
-    function getTestMapValueById(uint256 id) public view returns (uint256) {
-        return testMap[id];
-    }
+        // todo block timestamp and number are excessive in RootHistoryInfo
+        self.rootHistory.push(
+            RootHistoryInfo(self.root, _timestamp, _blockNumber)
+        );
 
-    function setTestMapValue(uint256 id, uint256 value) public {
-        testMap[id] = value;
+        addRootTransition(self, _timestamp, _blockNumber, 123);
     }
 
-    function add(uint256 _i, uint256 _v) public onlyWriter {
-        rootHistory.push(
-            RootHistoryInfo(root, uint64(block.timestamp), uint64(block.number))
-        );
+    /**
+     * @dev Add anode to the SMT
+     * @param _i Index of node
+     * @param _v Value of node
+     */
+    function add(
+        SmtData storage self,
+        uint256 _i,
+        uint256 _v
+    ) public {
         Node memory node = Node(NodeType.LEAF, 0, 0, _i, _v);
-        root = addLeaf(node, root, 0);
+        self.root = addLeaf(self, node, self.root, 0);
+
+        // todo block timestamp and number are excessive in RootHistoryInfo
+        self.rootHistory.push(
+            RootHistoryInfo(
+                self.root,
+                uint64(block.timestamp),
+                uint64(block.number)
+            )
+        );
+
+        addRootTransition(
+            self,
+            uint64(block.timestamp),
+            uint64(block.number),
+            123
+        );
+    }
+
+    /**
+     * ================ Changed Signature =================
+     */
+    function addRootTransition(
+        SmtData storage self,
+        uint64 _timestamp,
+        uint64 _blockNumber,
+        uint256 _replacedBy
+    ) internal {
+        self.rootTransitions[self.root].createdAtTimestamp = _timestamp;
+        self.rootTransitions[self.root].createdAtBlock = _blockNumber;
+        self.rootTransitions[self.root].root = self.root;
+        if (self.rootHistory.length > 1) {
+            self
+                .rootTransitions[self.rootHistory.length - 2]
+                .replacedAtTimestamp = _timestamp;
+            self
+                .rootTransitions[self.rootHistory.length - 2]
+                .replacedAtBlock = _blockNumber;
+            self.rootTransitions[self.rootHistory.length - 2].replacedBy = self
+                .root;
+        }
+        _replacedBy = 123;
```

Comment:
What is the logic for this assignment?

Reply:
The main goal here is to verify upgrading signature of function (added additional _replacedBy parameter)